kong/api/routes/upstreams.lua

```diff
@@ -0,0 +1,101 @@
+local crud = require "kong.api.crud_helpers"
+
+return {
+  ["/upstreams/"] = {
+    GET = function(self, dao_factory)
+      crud.paginated_set(self, dao_factory.upstreams)
+    end,
+
+    PUT = function(self, dao_factory)
+      crud.put(self.params, dao_factory.upstreams)
+    end,
+
+    POST = function(self, dao_factory, helpers)
+      crud.post(self.params, dao_factory.upstreams)
+    end
+  },
+
+  ["/upstreams/:name_or_id"] = {
+    before = function(self, dao_factory, helpers)
+      crud.find_upstream_by_name_or_id(self, dao_factory, helpers)
+    end,
+
+    GET = function(self, dao_factory, helpers)
+      return helpers.responses.send_HTTP_OK(self.upstream)
+    end,
+
+    PATCH = function(self, dao_factory)
+      crud.patch(self.params, dao_factory.upstreams, self.upstream)
+    end,
+
+    DELETE = function(self, dao_factory)
+      crud.delete(self.upstream, dao_factory.upstreams)
+    end
+  },
+
+  ["/upstreams/:name_or_id/targets/"] = {
+    before = function(self, dao_factory, helpers)
+      crud.find_upstream_by_name_or_id(self, dao_factory, helpers)
+      self.params.upstream_id = self.upstream.id
+    end,
+
+    GET = function(self, dao_factory)
+      crud.paginated_set(self, dao_factory.targets)
+    end,
+
+    POST = function(self, dao_factory, helpers)
+      local cleanup_factor = 10 -- when to cleanup; invalid-entries > (valid-ones * cleanup_factor)
+      
+      --cleaning up history, check if it's necessary...
+      local target_history = dao_factory.targets:find_all(
+            { upstream_id = self.params.upstream_id })
+      if target_history then --ignoring errors here, will be caught when posting below
+        -- sort the targets
+        for _,target in ipairs(target_history) do
+          target.order = target.created_at..":"..target.id
+        end
+        -- sort table in reverse order
+        table.sort(target_history, function(a,b) return a.order>b.order end)
+        -- do clean up
+        local cleaned = {} 
+        local delete = {}
+        for _, entry in ipairs(target_history) do
+          if cleaned[entry.target] then
+            -- we got a newer entry for this target than this, so this one can go
+            delete[#delete+1] = entry
+          else
+            -- haven't got this one, so this is the last one for this target
+            cleaned[entry.target] = true
+            cleaned[#cleaned+1] = entry
+            if entry.weight == 0 then
```

Comment:
yes. On the balancer object it gets deleted. But because we need individual Kong nodes to retain the same order of slots in the balancer, we must still record setting it to 0. This way we can replay changes on each node and end up with the same balancer config.